### PR TITLE
OSD-3393 handle 404s on S3 buckets

### PR DESCRIPTION
**Pull request by yithian**

### CODE REVIEW

**Comment by yithian** [comment]
I had to refactor _a lot_ of mock AWS test code, please let me know if you have any questions - I'd be more than happy to answer them or walk through things.

File: pkg/storage/s3/s3.go [hunk 1]
```diff
@@ -222,3 +191,56 @@ func generateBucketName(prefix string) string {
...
+	}
+
+	bucketinfo, err := ListBucketTags(s3Client, bucketlist.Buckets)
+	if err != nil {
+		return err
+	}
+
+	existingBucket := FindMatchingTags(bucketinfo, d.Config.InfraName)
+	if existingBucket != "" {
+		// HEAD the bucket. if it returns ~404 then label it as do-not-recover
```
Comment (cblecker) [inline pkg/storage/s3/s3.go:215]: This doesn't work. The problem is that `ListBucketsInRegion` and `ListBucketTags` return errors when the bucket doesn't exist, so we won't even get to this point.

Comment (yithian) [reply pkg/storage/s3/s3.go:215]: Well damn.

File: pkg/storage/s3/bucket_test.go [hunk 2]
```diff
@@ -54,6 +81,8 @@ func (c *mockAWSClient) GetAWSClientConfig() *aws.Config {
 
 // HeadBucket implements the HeadBucket method for mockAWSClient.
 // This mocks the AWS API response of having access to a single bucket named "testBucket".
+// NOTE "inconsistentBucket" is intentionally not recognized, in order to test how the
```
Comment (cblecker) [inline pkg/storage/s3/bucket_test.go:84]: Is this comment still accurate?

File: pkg/storage/s3/s3_test.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+		{
+			name:            "set recovered bucket name in instance status",
+			awsClient:       fakeClient,
+			bucketName:      "testBucket",
+			matchBucketName: true,
+		},
+		{
+			name:            "don't reclaim inaccessible bucket",
+			awsClient:       fakeInconsistentClient,
+			bucketName:      "inconsistentBucket",
```
Comment (cblecker) [inline pkg/storage/s3/s3_test.go:34]: where is the logic that deals with this differently?

File: pkg/storage/s3/s3_test.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+				t.Fatalf("got an unexpected error: %s", err)
+			}
+
+			// if the instace status' bucket name doesn't match the specified bucket name but is supposed to
+			if (instance.Status.StorageBucket.Name != tt.bucketName) && tt.matchBucketName {
+				t.Errorf("setInstanceBucketName() bucket name: %s, expected %s", instance.Status.StorageBucket.Name, tt.bucketName)
+			}
+
+			// if the instance status' bucket name matches the specified bucket name but isn't supposed to
+			if (instance.Status.StorageBucket.Name == tt.bucketName) && (tt.matchBucketName == false) {
```
Comment (cblecker) [inline pkg/storage/s3/s3_test.go:55]:
```suggestion
			if (instance.Status.StorageBucket.Name == tt.bucketName) && !tt.matchBucketName {
```

File: pkg/storage/s3/s3_test.go [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+		name            string
+		awsClient       *mockAWSClient
+		bucketName      string
+		matchBucketName bool
+	}{
+		{
+			name:            "set new bucket name in instance status",
+			awsClient:       fakeEmptyClient,
+			bucketName:      "",
+			matchBucketName: false,
```
Comment (cblecker) [inline pkg/storage/s3/s3_test.go:23]: this doesn't really check that it's set, it only checks that it's not a blank string. should we check that it `strings.HasPrefix(storageConstants.StorageBucketPrefix)`?

File: pkg/storage/s3/bucket_test.go (hunk 2 above)
Comment (yithian) [reply pkg/storage/s3/bucket_test.go:84]: It is still accurate, but it also applies to `GetBucketLocation()` and `GetBucketTagging()` as well, now and should be separated from the `HeadBucket()`-specific comment to make that clear~~er~~.

File: pkg/storage/s3/s3_test.go (hunk 3 above)
Comment (yithian) [reply pkg/storage/s3/s3_test.go:34]: The mock AWS client functions only "succeed" for a bucket named `testBucket`. Passing in a bucket named `inconsistentBucket` (or really, anything else), will cause the mock AWS client to return a NotFound awserr.

File: pkg/storage/s3/s3_test.go (hunk 5 above)
Comment (yithian) [reply pkg/storage/s3/s3_test.go:23]: I would assert that if `instance.Status.StorageBucket.Name` isn't an empty string then it is actually set.

That said however, your suggestion would definitely test the desired behavior more tightly.

Comment (cblecker) [reply pkg/storage/s3/s3_test.go:23]: Okay, yes, that's true. I guess where it makes a difference is in the other test cases.. are we *really* setting it to what we intend.

File: pkg/storage/s3/s3_test.go (hunk 3 above)
Comment (cblecker) [reply pkg/storage/s3/s3_test.go:34]: Is this true for your fakeInconsistentClient? You actually have a bucket named "inconsistentBucket". Or is this some magic in AWS' mocking?

Comment (yithian) [reply pkg/storage/s3/s3_test.go:34]: `fakeInconsistentClient` is a `mockAWSClient` that has been created with an array of (fake) `s3.Bucket`s with a single bucket named "inconsistentBucket". So when `fakeInconsistentClient.ListBuckets()` fires, it returns the list of buckets stored in the client - in this case, one named "inconsistentBucket". But none of the other functions (`GetBucketTagging()`, `GetBucketLocation()`, `HeadBucket()`, etc) return good results for a bucket named anything other than "testBucket".

So when the test has `fakeInconsistentClient` make the call to list its buckets, it returns a list of buckets containing "inconsistentBucket" and then when it tries to list the tags for it or get the its region or try to HEAD the bucket, those functions will return the `NotFound` awserr.

Comment (cblecker) [reply pkg/storage/s3/s3_test.go:34]: Ohhhh okay. I gotcha. Sounds good.

File: pkg/storage/s3/s3.go [hunk 6]
```diff
@@ -222,3 +191,55 @@ func generateBucketName(prefix string) string {
...
+			if (aerr.Code() == s3.ErrCodeNoSuchBucket) || (aerr.Code() == "NotFound") {
+				bucketLog.Info("Existing bucket is unavailable, disregarding it.")
+				return nil
+			} else {
+				return aerr
+			}
+		} else {
+			return err
+		}
+	}
```
Comment (cblecker) [inline pkg/storage/s3/s3.go:216]: We don't need to cast here, as we are doing it inside the function.

**Pull request merged**